### ISSUE TITLE
Xamarin SDK Error with self hosted parse-server

### ISSUE THREAD

**Issue by tomh4** [parse-community/Parse-SDK-dotNET#212]
Whenever I try to do any call to my Parse server I get an error. 
Whether it is a SaveAsync or a facbook login , it is always this error : 

>  {System.NullReferenceException: Object reference not set to an instance of an object
> 
>   at Parse.Internal.HttpClient+<>c__DisplayClass15.b__d (System.Threading.Tasks.Task1 t) [0x00093] in <filename unknown>:0 
>   at System.Threading.Tasks.ContinuationResultTaskFromResultTask2[TAntecedentResult,TResult].InnerInvoke () [0x00027] in /Users/builder/data/lanes/3236/ee215fc9/source/mono/external/referencesource/mscorlib/system/threading/Tasks/TaskContinuation.cs:207 

and so on and so forth.
I did initialize it correctly to my server, and I can confirm the server is working as I can do stuff with cURL.

I am using the Nuget version

**Comment by thewizster** [comment]
Please post the parse-server and SDK versions you are using. You said Xamarin SDK but I assume you are working with the Parse Client SDK's to communicate with parse-server?

And, would you give an example snippet of code? That way we can try and reproduce the error? Not enough info to solve a null ref exception...

**Comment by tomh4** [comment]
Hello, 
Thank you for getting back.
So here is what I do :
Im using Xamarin and the Parse NUGET so the 1.7 SDK together with parse server 2.2.10
I initialize the server
ParseClient.Initialize(new ParseClient.Configuration
{
ApplicationId = "MYAPPID",
Server = "MYSERVER"
});
ParseFacebookUtils.Initialize("MYFBID");

I removed the sensitive data now.
Next I try anything and it gives an error 
Wheter I I I facebook login

try
{
await ParseFacebookUtils.LogInAsync(id, accessToken, expiryDate);
return true;
}
catch (System.Exception e)
{
Console.Write(e);
return false;
}


Or simply try to save an object, always the error above

**Comment by thewizster** [comment]
Can you verify that you have a trailing slash in your server URL?

`Server = "http://example.com/parse/"`

Initializing your ParseClient should look something like this:
```
ParseClient.Initialize(new ParseClient.Configuration
{
ApplicationId = "MYAPPID",
Server = "http://example.com/parse/"
});
```

**Comment by tomh4** [comment]
Hello,
yes I can confirm it is ending with a trailing /
However it uses https, might that be a problem ?

**Comment by thewizster** [comment]
Just a few questions to try and understand your setup better:

1) Can you access the parse-server via the REST API or other SDK and verify that your server is functioning properly?

2) Do you have the `dotNetKey` parameter set in your parse-server config? If so, just remove it. All you need is `appId` param for parse-server.

Here is the basic server config:
```
var api = new ParseServer({
  databaseURI: 'mongodb://your.mongo.uri',
  cloud: './cloud/main.js',
  appId: 'myAppId',
  fileKey: 'myFileKey',
  masterKey: 'mySecretMasterKey',
  push: { ... }, // See the Push wiki page
  filesAdapter: ...,
});
```

**Comment by tomh4** [comment]
Hello,
yes i have the parse dashboard runnning and it shows the server. 
Hitting this in the terminal : 
`curl -X POST -H "X-Parse-Application-Id: com.app.id" -H "Content-Type: application/json" -d '{"score":1337,"playerName":"Sean Plott","cheatMode":false}' https://website/parse/classes/GameScore`
returns : 
`{"objectId":"Z8HHMmyjBc","createdAt":"2016-06-09T20:19:26.630Z"}`
and allows me to see the created object in the dashboard.

my sevrer config is as follows : 
I am using pm2 to start a service trough my file called ecosystem.json containing : 
`{
  "apps" : [{
    "name"        : "parse-wrapper",
    "script"      : "/usr/bin/parse-server",
    "watch"       : true,
    "merge_logs"  : true,
    "cwd"         : "/home/parse",
    "env": {
      "PARSE_SERVER_CLOUD_CODE_MAIN": "/home/parse/cloud/main.js",
      "PARSE_SERVER_DATABASE_URI": "Hidden",
      "PARSE_SERVER_APPLICATION_ID": "Hidden",
      "PARSE_SERVER_MASTER_KEY": "Hidden",
      "PARSE_SERVER_FACEBOOK_APP_IDS": "Hidden",
    }
  }]
}`

**Comment by thewizster** [comment]
Thanks for sharing. I'm going to setup my local dev env with a similar setup to yours and test on my end.

I did notice a spare comma in your JSON at the end of the `PARSE_SERVER_FACEBOOK_APP_IDS` line. My editor didn't like it. That may just be a typo on github but I just wanted to point it out. I have removed it and re-posted the JSON here just for clarity sake.

```
{
  "apps": [
    {
      "name": "parse-wrapper",
      "script": "/usr/bin/parse-server",
      "watch": true,
      "merge_logs": true,
      "cwd": "/home/parse",
      "env": {
        "PARSE_SERVER_CLOUD_CODE_MAIN": "/home/parse/cloud/main.js",
        "PARSE_SERVER_DATABASE_URI": "Hidden",
        "PARSE_SERVER_APPLICATION_ID": "Hidden",
        "PARSE_SERVER_MASTER_KEY": "Hidden",
        "PARSE_SERVER_FACEBOOK_APP_IDS": "Hidden"
      }
    }
  ]
}
```

**Comment by tomh4** [comment]
Thanks for taking your time.
Ill check if the comma is there or not, but the server is working so im not sure why it should have an impact on the sdk:/ ill check it out

**Comment by tomh4** [comment]
Any news ? dont want to push but im sitting here for weeks not being able to continue work....

**Comment by thewizster** [comment]
Note: I had no previous experience with PM2 prior to this and I wasn't working on my Mac at the time. But here was my experience.

I setup PM2 (v1.1.3) on my Win10 dev machine and ran parse-server (2.2.13) locally connecting to a local dev mongodb server. Here are my results:

1) Ran parse-server using `pm2 start index.js` parse-server ran normally and I was able to connect using a locally hosted parse-dashboard (v1.0.14) instance and browse/add data. I also was able to use Postman to interact with the parse-server and add data. No issues.

2) Created a `process.json` file, similar to your `ecosystem.json`, and ran `pm2 start process.json` and again the results were same as number 1. All good!

3) One and two were both in PM2 fork mode so I wanted to try and get cluster mode working. I was able to configure cluster mode. I got a 3 cluster parse-server running BUT, I was never able to get parse-server to work under cluster mode.

Since parse-server is still a work in progress and based on my experience I would drop PM2 and just get back to running a plain vanilla parse-server for the moment. Then work back into it until you find the problem. I can assure you that the .NET SDK does indeed work with parse-server. I am using it.

@tomh4 I'm just a parse user like you my friend! I'm just volunteering my efforts to try and help out. I am at the end of my available time for this issue. Sorry I wasn't able to help more. Good luck!

**Comment by tomh4** [comment]
First of all thanks for your time. My comment was not intended to be offensive, i am thankfull that you are around because I think noone else is...

I am not sure what you mean by running a plain vanilla parse-server
I use pm2 because it is easy to use and allows me to have the service restarted automatically in case of a restart of the servrer.
I think i will just use the rest api...

**Comment by thewizster** [comment]
No worries...

I mean for you to setup a fresh parse-server like shown in the parse-server example code and then run your client code against it.

Also, If you are using Visual Studio or Xamarin Studio you can download this example parse client console app and test your parse-server environment. It uses the current NuGet of Parse .NET SDK.

https://github.com/thewizster/parse-server-console

**PLEASE NOTE:** This console app **WILL** write data to your parse-server so be sure you don't have a Hello class or a World class in your target server that has important data saved. The console app will write data to those two parse-server classes.

**Comment by tomh4** [comment]
So did you run it within express? The server I mean

**Comment by thewizster** [comment]
Yes. Parse Server + Express

```
var express = require('express');
var ParseServer = require('parse-server').ParseServer;
var app = express();

var api = new ParseServer({
  databaseURI: 'mongodb://localhost:27017/dev', // Connection string for your MongoDB database
  cloud: '/home/myApp/cloud/main.js', // Absolute path to your Cloud Code
  appId: 'myAppId',
  masterKey: 'myMasterKey', // Keep this key secret!
  fileKey: 'optionalFileKey',
  serverURL: 'http://localhost:1337/parse' // Don't forget to change to https if needed
});

// Serve the Parse API on the /parse URL prefix
app.use('/parse', api);

app.listen(1337, function() {
  console.log('parse-server-example running on port 1337.');
});
```

**Comment by couchworx** [comment]
Hello,
I'm having the same problem. I've setup parse-server on AWS / Elastic Beanstalk connected it to my migrated database on mLab. Calls via curl / http work just fine. Using Parse Dashboard works also. Using Parse Server 2.2.13 and the SDK version 1.7.0 (Xamarin.iOS) throws the same exception doing a simple call:
`await ParseUser.LogInAsync(username, password);`

`Object reference not set to an instance of an object`
`  at Parse.Internal.HttpClient+<>c__DisplayClass15.<ExecuteAsync>b__d`

**Comment by thewizster** [comment]
Please post the code you are running when calling `LogInAsync()`

**Comment by thewizster** [comment]
@couchworx Please post the code you are running when calling `LogInAsync()` that will help understand your issue better.

Also, I have a small console app that posts some test data using the .Net client SDK. You could run that against your server and make sure it's working. Get the project here:
https://github.com/thewizster/parse-server-console

**Comment by thewizster** [comment]
Here is some more sample code that might be of use for this discussion.

https://github.com/thewizster/hello-xamarinforms-ios-parse

Basic Xamarin iOS application that saves data to parse-server. Just change your appId and server url then run the app. If your server is running and properly configured It will save some test data to your parse-server.

**Comment by couchworx** [comment]
@thewizster Thanks for the quick response. I've just tried the basic Xamarin iOS application. Unfortunately it shows the same issue. If you want to try, I've temporary changed the appId to "47114712", the server URL is "http://puzoh-env.us-west-2.elasticbeanstalk.com:1337/parse/"

![simulator screen shot jun 16 2016 9 32 12 am](https://cloud.githubusercontent.com/assets/4161797/16124762/37c2e492-33a5-11e6-8106-02f85148de85.png)

**Comment by couchworx** [comment]
Posting data via curl works nicely. Feel free to try yourself, the server is for testing. ;)

`curl -X POST \
  -H "X-Parse-Application-Id: 47114712" \
  -H "Content-Type: application/json" \
  -d '{"score":1337,"playerName":"Sean Plott","cheatMode":false}' \
  http://puzoh-env.us-west-2.elasticbeanstalk.com/parse/classes/GameScore`

**Comment by thewizster** [comment]
@couchworx You are using `http://puzoh-env.us-west-2.elasticbeanstalk.com:1337/parse/` as the server URL in your app which has  http port 1337 assigned. There is no site at this address.

You need to use: `http://puzoh-env.us-west-2.elasticbeanstalk.com/parse/`

**Note:** I have removed `:1337` from the url since your server is operating on default port 80

Your curl works because you are posting to default http port 80.

After using the updated URL I ran my app against your server and it worked.

**Comment by couchworx** [comment]
@thewizster Perfect, don't know how I missed that. Everything seems to be working fine. Thanks for your help.

**Comment by tomh4** [comment]
So i have tried your console app and it works fine with my server. however the xamarin app still doesnt work ... what version of the sdk is your console app using ?

**Comment by thewizster** [comment]
@tomh4  It's using the NuGet package v1.7.0
https://www.nuget.org/packages/parse/1.7.0

Take a look at the packages.config file.
https://github.com/thewizster/parse-server-console/blob/master/ParseServerConsole/packages.config

**Comment by tomh4** [comment]
Same package I am using...                                                                                                                                                                                                                                                                                                                                                     Von: Raymond BradyGesendet: Samstag, 18. Juni 2016 18:30An: ParsePlatform/Parse-SDK-dotNETAntwort an: ParsePlatform/Parse-SDK-dotNETCc: tomh4; MentionBetreff: Re: [ParsePlatform/Parse-SDK-dotNET] Xamarin SDK Error with self hosted parse-server (#212)@tomh4  It's using the NuGet package v1.7.0
https://www.nuget.org/packages/parse/1.7.0

Take a look at the packages.config file.
https://github.com/thewizster/parse-server-console/blob/master/ParseServerConsole/packages.config

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by tomh4** [comment]
My packages.config had 
<package id="Parse" version="1.7.0" targetFramework="monodroid60" />
  
but i changed it to 
<package id="Parse" version="1.7.0" targetFramework="net461" />
  
and still not working

**Comment by rfrailemartinez** [comment]
I have the same problem.

But there is something special I did:
- Test 1: App is connected to 4G or normal Wifi: I have same error as tomh4
- Test 2: I put a intermediate proxy -> Fiddler and I have installed its own root certificate on android device for https.The apps works perfect.

Because of that, I think the problem maybe comes from:
- redirection in server nginx (https:://dns to http://ip:1337), and parse httpclient don´t allow this kind of redirection
- https certificate from letsencryt is considerated by android like not valid

I have mounted a digital ocean droplet with parse server and parse dashboard, both of them started from pm2 configuration ans exposed with nginx. At this moment and with this error I can not migrate my apps.

thanks
